### ISSUE TITLE
Network interface names restricted to RFC 1123 DNS label syntax

### ISSUE THREAD

**Issue by TheDiveO** [intel/multus-cni#213]
At this time, the multus Kubernetes client checks in [`parsePodNetworkObjectName()`](https://github.com/intel/multus-cni/blob/master/k8sclient/k8sclient.go#L181) the network **interface name** against DNS label syntax conformance. While in Kubernetes namespace and resource name identifiers are limited to DNS label cornformance, the names of network interfaces -- see: [network interface names on Linux](https://unix.stackexchange.com/questions/451368/allowed-chars-in-linux-network-interface-names) -- are not limited at all to DNS label syntax. In fact, a lot of CNI plugins -- not least flannel -- use Linux network interface names that are incompatible with DNS label rules, such as `flannel.1`. At least to my admittedly limited understanding the names of network interfaces are not related to Kubernetes/Multus resources, so the DNS label restrictions should not apply then?

If I'm not missing places where a (container) network interface might be used to reference a Kubernetes resource, then the check linked to above is an unnecessary restriction. Can it be lifted then?

A side effect of lifting the restriction would then be that the "split" checks need to be made of the new situation:

- the optional "@" network interface name is best split off **before** splitting of the "/" namespace: the rationale here being that a network interface name itself may contain "/". The current implementation would then trip off.
- the "@" split must occur only on the first "@".
- the "/" split must only occur after the "@" split, in order to not make it trip over "/" in network interface names.

**Comment by dougbtv** [comment]
Thanks for the input @TheDiveO ! 

Two parts -- I believe this is only used when you specify the interface name directly. Otherwise, you'll get incremental names, e.g. `net0, net1, net2, ...netN`.

I believe this may have been in place due to earlier iterations of the [Network Plumbing Working Group's de-facto specification](https://github.com/K8sNetworkPlumbingWG/multi-net-spec/blob/master/%5Bv1%5D%20Kubernetes%20Network%20Custom%20Resource%20Definition%20De-facto%20Standard.md), see especially `3.3 NetworkAttachmentDefinition Object Naming Rules` -- where this convention still stands that we used DNS 1123 labels. I think at some point the spec stated that the interface names must also follow this convention, however -- I don't believe that to be the case anymore, seen here:

> 4.1.2.1.5 "interface"
> This optional key with value of type string requires the implementation to use the given name for the pod interface resulting from this network attachment. This keys value must be a valid Linux kernel interface name

So, likely we should probably conform to the spec. 

Do you have some use cases where you need to specify an interface name more loosely? Would love to hear about it. The only thing I currently like about how it's implemented (while it lacks in conformance to the specification) is just that it's fairly easy to validate the data the user passed without allowing too much garbage, and makes it easy to parse.

**Comment by TheDiveO** [comment]
If I didn't misunderstood section 3.3, then we're talking solely about the network attachment resources in Kubernetes, but not Linux kernel network interface names in this context. Of course, this will always hold, as it is set by Kubernetes' architecture, so it's DNS label syntax.

At this time I luckily don't have the need for more arcane Linux network interface names, especially not within containers. I just noticed the questionable check while trying to understand multus" inner workings. As for Linux network interface names, I've linked to what I found out while working on some project trying to ensure that I can properly process nif names returned via RTNETLINK and trying to figure out iproute2's slightly arcane nif name adornments, such as `eth0@if42`.
